parse_node/parse_node_with_comments.ts

```diff
@@ -0,0 +1,180 @@
+import ts from "typescript"
+import {
+  ParseNodeType,
+  parseNodeWithoutComments,
+  ParseState,
+} from "../parse_node"
+import { Test } from "../tests/test"
+
+const getBackwardsClosestNewLinesCount = (text: string, pos: number) => {
+  let count = 0
+
+  for (let i = pos - 1; i >= 0; i--) {
+    const char = text.charAt(i)
+    if (char === "\t" || char === " ") {
+      continue
+    }
+
+    if (char === "\n") {
+      count++
+      continue
+    }
+
+    break
+  }
+
+  return count
+}
+
+export const parseNodeWithComments = (
+  node: ts.Node,
+  props: ParseState
+): ParseNodeType => {
+  if (ts.isSourceFile(node)) {
+    return parseNodeWithoutComments(node, props)
+  }
+
+  let leadingComments =
+    ts
+      .getLeadingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let tarilingComments =
+    ts
+      .getTrailingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart() + node.getFullWidth()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let commentStackUnwind = props.commentsStack?.length ?? 0
+
```

Comment:
For my own edification, maybe a comment here such as:

`// gather up all comments`